### ISSUE TITLE
Dogfood

### ISSUE THREAD

**Issue by timm** [researchart/rose6icse#24]
https://github.com/researchart/rose6icse/tree/master/submissions/available/dogfood
https://github.com/researchart/rose6icse/tree/master/submissions/reusable/dogfood

email: dongjie_cdj@163.com github id: midwinter1993

**Comment by minkull** [comment]
Dongjie Chen
State Key Lab for Novel Software
Technology, Nanjing University
Nanjing, China
Department of Computer Science and
Technology, Nanjing University
Nanjing, China
dongjie_cdj@163.com
Yanyan Jiang
State Key Lab for Novel Software
Technology, Nanjing University
Nanjing, China
Department of Computer Science and
Technology, Nanjing University
Nanjing, China
jyy@nju.edu.cn
Chang Xu
State Key Lab for Novel Software
Technology, Nanjing University
Nanjing, China
Department of Computer Science and
Technology, Nanjing University
Nanjing, China
changxu@nju.edu.cn
Xiaoxing Ma
State Key Lab for Novel Software
Technology, Nanjing University
Nanjing, China
Department of Computer Science and
Technology, Nanjing University
Nanjing, China
xxm@nju.edu.cn
Jian Lu
State Key Lab for Novel Software
Technology, Nanjing University
Nanjing, China
Department of Computer Science and
Technology, Nanjing University
Nanjing, China
lj@nju.edu.cn

**Comment by dmz-peerreview** [comment]
Please check first link, it seems to be broken / not available

**Comment by minkull** [comment]
There does not seem to be any folder for this submission under "available". I suspect that this submission is only for "reusable", and that the inclusion of "available" there may have been a typo.

@midwinter1993 could you confirm this?

**Comment by midwinter1993** [comment]
Hi, @minkull 
I'd like to apply for the `available` badge as well. Sorry that I did not put the artifact in the subfolder. 
I have made a pull request which contains the artifact in the subfolder.

**Comment by minkull** [comment]
@midwinter1993 Ok, no worries. I've just accepted the pull request.

**Comment by happyanonymous** [comment]
@grumpyreviewer have you had any success using the environment shared via docker? I'm trying to reproduce the steps on the https://github.com/researchart/rose6icse/tree/master/submissions/reusable/dogfood, but it seems that I don't have the necessary infrastructure in my docker container (e.g., I have no ./ctrl directory)

**Comment by midwinter1993** [comment]
Hi @happyanonymous,
I pulled the docker image and tested it again.

All artifacts are under directory `/dogfood` and the `ctrl` is a script whose absolute path is `/dogfood/workspace/ctrl`.

PS: As mentioned in the `INSTALL.md`: "All the following commands are run under directory `/dogfood/workspace/`".

**Comment by danielcalencar** [comment]
Thanks @midwinter1993 , I think my initial feedback would be to mention it in the "Reproduce results in the paper" as well, not only in the INSTALL.md. 

My other question is: although you provide a Docker, would I necessarily have to run it on a Ubuntu machine? Because that's what the text in the "Obtain the artifact" section implies.

B

**Comment by happyanonymous** [comment]
Thanks @midwinter1993 , I think my initial feedback would be to mention it in the "Reproduce results in the paper" as well, not only in the INSTALL.md.

My other question is: although you provide a Docker, would I necessarily have to run it on a Ubuntu machine? Because that's what the text in the "Obtain the artifact" section implies.

**Comment by happyanonymous** [comment]
As of now, I can only indicate *Available*, but not *Reusable*. I've tried two settings:

- on a Windows 10 machine with Docker version 18.03.0-ce, build 0520e24302: I can't reproduce the results in the paper since I can't run the command `docker run -it --rm --device /dev/kvm dogfood:v1`

- I've also tested on a VM containing Ubuntu 18.04 and face the same issue as above.

I've tried to run the following command instead: `docker run -ir --rm midwinter/doogfood:v1`. While I'm able to access the Docker container I can't install qemu within the Docker container.

PS: Although the authors mention that the experiments should be ran on a Ubuntu machine because it uses KVM, it also renders the Docker effectively unusable. Therefore, I'm not sure if I can recommend the *Reusable* badge

**Comment by midwinter1993** [comment]
Hi @happyanonymous,

If you run on Windows or VM, QEMU is unable to use the KVM and it will run slowly.

You can delete the KVM option in the file `/dogfood/workspace/mngr-tools/config.sh`

```bash
# KVM_OPT="-enable-kvm -smp 2"
```

**Comment by happyanonymous** [comment]
Hi @midwinter1993 

Thank you for letting us know. Again, I think it is important to have it clearly described on the page of your artifact.

Not sure if you missed it in my previous comment, but I also face a bug when trying to install qemu in the docker. Is it possible for you to provide a Docker containing qemu (and other necessary tools ) already installed? I think the whole point of making the Docker available is to provide a ready-to-go environment with minimum setup possible --- so that we can concentrate on running the commands and reproducing your results only.

Regards,

**Comment by midwinter1993** [comment]
Hi @happyanonymous,

The Docker image contains all dependencies including QEMU.
So, there is no need to install it again.

Sorry that I found an issue:
Change the `TOP` variable in the file `/dogfood/workspace/mngr-tools/config.sh` to

```bash
TOP="/dogfood/workspace"
```

I have uploaded a gif to show how to start the QEMU.
[Gif here](https://drive.google.com/file/d/1NItOEx88Z30HKbjlnSik_oTwJa7T4ceM/view?usp=sharing)

**Comment by happyanonymous** [comment]
Hi @midwinter1993 ,

Thanks for letting us know how to fix the bug in your Docker setup. I'll get back to you regarding this soon.

Regards,

**Comment by timm** [comment]
dear @midwinter1993, thank you so much for your submission and your responsiveness.

@happyanonymous : thanks for your detailed interaction. what is the current status? should @grumpyreviewer go ahead? or should we require more changes from @midwinter1993 first?

please advise

**Comment by happyanonymous** [comment]
I am still getting a bug after following the gif provided by @midwinter1993 
![image](https://user-images.githubusercontent.com/44096371/74797574-489ce980-5330-11ea-89f7-77194271611f.png)

I'll stick with *Available* only. I think @grumpyreviewer  should go ahead.

Cheers,

**Comment by midwinter1993** [comment]
Hi @happyanonymous,

Did you follow the `INSTALL.md` to decompress the file `objs.tar.gz` first?

```bash
tar -xzvf objs.tar.gz
```

The extracted directory `objs` will contain the `qcow2` file.

```
objs
├── debian-image
│   ├── create-fs-image.sh
│   ├── create-image.sh
│   ├── stretch.id_rsa
│   ├── stretch.id_rsa.pub
│   ├── stretch.img
│   └── stretch.img.qcow2
├── linux-5.1.3
│   ├── arch
│   │   ├── x86
│   │   │   └── boot
│   │   │       └── bzImage
│   │   └── x86_64
│   │       └── boot
│   │           └── bzImage -> ../../x86/boot/bzImage
│   └── vmlinux
├── qemu-hda.img
└── qemu-hda.img.qcow2
```

If that does not work, could you please attach all the steps (commands) you executed?
I will figure out the issue as soon as possible. :-P

**Comment by timm** [comment]
Current status:

- 1 vote for "available"
- awaiting review from  @grumpyreviewer

**Comment by grumpyreviewer** [comment]
TL;DR: _available_ ok, _reusable_ may need some more work (the documentation is not detailed enough to allow repurposing; several errors when using more advanced features). 

I have some concerns about the functionality, since I could not run all the commands in INSTALL.md due to a few errors. But I reckon the specific artifact relies on a number of hw and os features that I might have failed to configure correctly in my environment. Open to discuss this point if also @happyanonymous has concerns in this direction.

The docker command needed some fixes, as pointed out by @happyanonymous 
docker run -it --rm --device /dev/kvm midwinter/doogfood:v1

Running on Mac was a lot of troubles, while on a bare metal Ubuntu with kvm the container ran, but with a few issues:

1) hard-coded paths:
./ctrl start
qemu-system-x86_64: -hda /home/dongjie/Documents/dogfood/workspace/objs/debian-image/stretch.img.qcow2: Could not open '/home/dongjie/Documents/dogfood/workspace/objs/debian-image/stretch.img.qcow2': No such file or directory

due to $TOP in config.sh:
TOP="/home/dongjie/Documents/dogfood/workspace"

2) Should install telnet in the docker
telnet: command not found
after installing telnet and running `telnet localhost 45454`:
EMU 2.11.1 monitor - type 'help' for more information
(qemu) info snapshots
There is no snapshot available.

3) Kernel compilation from scratch did not work for me
./build kernel
...
Error: kernelrelease not valid - run 'make prepare' to update it
...
Compiler lacks asm-goto support.
arch/x86/Makefile:301: recipe for target 'checkbin' failed
make[1]: *** [checkbin] Error 1

I tried to add a make prepare, but I did not manage to solve the issue.

**Comment by midwinter1993** [comment]
Hi @grumpyreviewer and @happyanonymous,

Thank you for pointing out these issues.

TL;DR: I have updated a docker image containing everything.

```bash
docker pull midwinter/dogfood:v3
```

1. I did not install tools to build kernels from scratch in the docker container and
compressed files to reduce the image size, which seems to hinder the reuse.
Now, in `midwinter/dogfood:v3`, I installed all the building tools and extracted all the files.
Just to run the `dogfood` (`./ctrl start` and `./ctrl init`), and building kernel from scratch is easy by `./build kernel`.

2. After running `./ctrl start`, you should run `./ctrl init` in another terminal (`tmux` is installed); then, you will see snapshots via telnet. @grumpyreviewer

**Comment by timm** [comment]
@happyanonymous  
@grumpyreviewer  

- currently we have agreed on "available"

- 2 hours ago, the authors pushed a new version. please check for "reusable" (i.e. it runs) and advise your results below

thank you!

**Comment by happyanonymous** [comment]
So, I am keeping the vote for *Available* as I am getting an error when pulling the new docker version:

![image](https://user-images.githubusercontent.com/44096371/74991639-588d0880-54ab-11ea-894e-74b2b0fa92e3.png)

Is it something that I have to do on my side? I feel like I have plenty of space though... Or it is something related to the docker image?

**Comment by midwinter1993** [comment]
Hi @happyanonymous,

Maybe there is no enough space for docker image storage in your computer (not the space of disk).
Try to prune unused images. [here](https://stackoverflow.com/questions/30604846/docker-error-no-space-left-on-device)

I have tested the image again, it seems okay in my laptop.
![image](https://user-images.githubusercontent.com/7164461/75004242-1ffc1900-5430-11ea-89e0-339c4c07dbbc.png)

**Comment by happyanonymous** [comment]
Hi @midwinter1993,

I got a problem when building the kernel (`./build kernel`)

![image](https://user-images.githubusercontent.com/44096371/75008562-37dda680-54dd-11ea-8e54-a04cdf03cf60.png)

**Comment by happyanonymous** [comment]
When commenting the KVM option in the `config.sh', I get the following error:
![image](https://user-images.githubusercontent.com/44096371/75009240-0fef4280-54df-11ea-89de-a60ebb970ee1.png)

Any advice?
Cheers,

**Comment by midwinter1993** [comment]
Maybe there is no enough RAM in your computer.

How about to change the parameter in the file `/dogfood/workspace/mngr-tools/mngr-lib.sh:80`
`-m 4G` to a small value, e.g., `-m 1G`?

**Comment by timm** [comment]
currently we are converging   on "available".

**Comment by happyanonymous** [comment]
Thanks for your help @midwinter1993 ,

I could finally feel like a happy-reviewer because I could reproduce your results now. Perhaps my feedback would be to add a troubleshooting section to your documentation, since we had to go over many things to make things work.

Also, I do not see an unique identifier associated to your artifact (e.g., a DOI). If those things are addressed I recommend *reusable*.

Cheers,

**Comment by timm** [comment]
currently we are converging on "available".

**Comment by midwinter1993** [comment]
Hi @happyanonymous,

Thank you for your advice!

I have revised the documentation to make the install and running more clear, and also added a troubleshooting section.
A pull request is made; if it has not been merged, you can check it in my [repo](https://github.com/midwinter1993/dogfood).

A DOI has been put in the `available` subfolder, but not in the `reusable` subfolder.
I have added it in both folders.

**Comment by minkull** [comment]
Many thanks everyone for the discussions!

Available awarded.
